proof/refine/ARM/Finalise_R.thy

```diff
@@ -2329,12 +2329,162 @@ lemma replyUnlink_invs'[wp]:
+
+lemma replyUnlink_ko_at'_reply_sp:
+  "\<lbrace>ko_at' reply replyPtr\<rbrace>
+   replyUnlink replyPtr tcbPtr
+   \<lbrace>\<lambda>rv. ko_at' (reply\<lparr>replyTCB := None\<rparr>) replyPtr\<rbrace>"
+  apply (simp add: replyUnlink_def)
+  by (wpsimp wp: setReplyTCB_ko_at'_reply_sp hoare_vcg_imp_lift simp: replyUnlink_assertion_def getReplyTCB_def)
+
+crunches tcbSchedDequeue, schedContextDonate
+  for ko_at'_reply[wp]: "ko_at' (reply::Structures_H.reply) replyPtr"
+
+lemma replyPop_list_refs_of_replies'[wp]:
+  "replyPop replyPtr tcbPtr \<lbrace>\<lambda>s. sym_refs (list_refs_of_replies' s)\<rbrace>"
+  unfolding replyPop_def
+  apply (rule hoare_seq_ext[OF _ get_reply_sp'])
+  apply (rename_tac reply)
+  apply (rule hoare_seq_ext_skip, solves \<open>wpsimp\<close>, simp?)+
+  apply (rule hoare_seq_ext[rotated])
+   apply (rule hoare_conj,
+          rule replyUnlink_list_refs_of_replies',
+          rule replyUnlink_ko_at'_reply_sp)
+  apply simp
+  apply (rule hoare_when_cases, simp)
+  apply (rule hoare_seq_ext[OF _ assert_sp])
+  apply (rule hoare_seq_ext_skip, solves \<open>wpsimp simp: comp_def\<close>, simp?)+
+  apply (rule hoare_seq_ext[OF cleanReply_list_refs_of_replies'])
+  apply (rule hoare_when_cases)
+   apply (elim conjE delta_sym_refs
+          ; clarsimp simp: isHead_def split: if_splits)
+   apply (erule in_list_refs_of_replies'_E
+          ; clarsimp simp: obj_at'_def projectKOs)
+  apply (wpsimp simp: isHeadSome)
+  apply (rename_tac prevReplyPtr prevReply scPtr)
+  apply (frule ko_at'_replies_of')
+  apply (frule (3) sym_refs_replyNext)
+  apply (erule delta_sym_refs
+         ; clarsimp simp: list_refs_of_reply'_def in_get_refs replyPrev_list_refs_of_replies'
+                   split: if_splits)
+   apply (rule conjI | clarsimp simp: obj_at'_def projectKOs | erule in_list_refs_of_replies'_E)+
+  done
+
+lemma updateReply_list_refs_of_replies':
+  assumes "\<And>reply. list_refs_of_reply' (f reply) = k (list_refs_of_reply' reply)"
+  shows "\<lbrace>\<lambda>s. P (map_set (heap_upd k replyPtr (list_refs_of_replies_opt' s)))\<rbrace>
+         updateReply replyPtr f
+         \<lbrace>\<lambda>rv s. P (list_refs_of_replies' s)\<rbrace>"
+  apply (wpsimp simp: updateReply_def heap_upd_def assms)
+  apply (erule subst[rotated])
+  apply (rule arg_cong[where f=P, OF all_ext], clarsimp)
+  by (clarsimp simp: map_set_def ko_at'_replies_of' opt_map_left_Some)
+
+lemma bar:
+  "list_refs_of_reply' (reply \<lparr> replyPrev := Nothing \<rparr>)
+   = Set.filter (\<lambda>ref. snd ref \<noteq> ReplyPrev) (list_refs_of_reply' reply)"
+  apply (simp add: Set.filter_def list_refs_of_reply'_def get_refs_def split: option.splits)
+  sorry
+
+lemma cleanReply_list_refs_of_replies':
+  "\<lbrace>\<lambda>s. P (map_set (heap_upd k replyPtr (list_refs_of_replies_opt' s)))\<rbrace>
+   cleanReply replyPtr
+   \<lbrace>\<lambda>rv s. P (list_refs_of_replies' s)\<rbrace>"
```

Comment:
Can you prove this lemma using `updateReply_list_refs_of_replies’` above, instead of unfolding `updateReply`"